pkg/controller/node/cidr_allocator.go

```diff
@@ -0,0 +1,144 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package node
+
+import (
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"math/big"
+	"math/rand"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/golang/glog"
+)
+
+var errCIDRRangeNoCIDRsRemaining = errors.New("CIDR allocation failed; there are no remaining CIDRs left to allocate in the accepted range")
+
+// CIDRAllocator is an interface implemented by things that know how to allocate/occupy/recycle CIDR for nodes.
+type CIDRAllocator interface {
+	AllocateNext() (*net.IPNet, error)
+	Occupy(*net.IPNet)
+	Release(*net.IPNet)
+}
+
+type rangeAllocator struct {
+	clusterCIDR    *net.IPNet
+	subNetMaskSize int
+	cidrs          chan *net.IPNet
+	used           big.Int
+	lock           sync.Mutex
+	rand           *rand.Rand
+}
+
+// NewCIDRRangeAllocator returns a CIDRAllocator to allocate CIDR for node
+// Caller must ensure subNetMaskSize is not less than cluster CIDR mask size.
+func NewCIDRRangeAllocator(clusterCIDR *net.IPNet, subNetMaskSize int) CIDRAllocator {
+	ra := &rangeAllocator{
+		clusterCIDR:    clusterCIDR,
+		subNetMaskSize: subNetMaskSize,
+		rand:           rand.New(rand.NewSource(time.Now().UnixNano())),
+	}
+	return ra
+}
+
+func (r *rangeAllocator) AllocateNext() (cidr *net.IPNet, err error) {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+
+	clusterIP := r.clusterCIDR.IP.To4()
+	clusterMask := r.clusterCIDR.Mask
+	clusterMaskSize, _ := clusterMask.Size()
+
+	var maxCIDRs int
+	maxCIDRs = 1 << uint32(r.subNetMaskSize-clusterMaskSize)
+
+	nextUnused, err := func() (int, error) {
+		j := r.rand.Intn(maxCIDRs)
+		if r.used.Bit(j) == 0 {
+			return j, nil
+		}
+		// search sequentially
+		for i := j + 1; i < maxCIDRs; i++ {
+			if r.used.Bit(i) == 0 {
+				return i, nil
+			}
+		}
+		for i := 0; i < j; i++ {
+			if r.used.Bit(i) == 0 {
+				return i, nil
+			}
+		}
+		return -1, fmt.Errorf("not found")
```

Comment:
Just return `errCIDRRangeNoCIDRsRemaining` here instead of constructing an error you know you're going to throw away.

Then below, you can have:
```
if err != nil {
	return nil, err
}